### PR TITLE
Support for creating/updating secrets

### PR DESCRIPTION
**Pull request by omrirh**
- update pull-secret with new content using it's data as environment variable

### CODE REVIEW

**Comment by omrirh** [comment]
@rnetser @myakove this code is compatible for getting the pull-secret via a path; let me know if that should be changed

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_utilities/infra.py [hunk 1]
```diff
@@ -288,3 +292,18 @@ def cluster_resource(base_class):
...
+    with open(os.environ["PS_PATH"]) as ps_dict:
+        ps_dict_encoded = base64.b64encode(ps_dict.read().encode("ascii")).decode(
+            "utf-8"
+        )
+    ps_content = {".dockerconfigjson": ps_dict_encoded}
+
+    with cluster_resource(Secret)(
+        name="pull-secret", namespace="openshift-config"
+    ) as pull_secret:
+        ResourceEditor(patches={pull_secret: {"data": ps_content}}).update()
```
[the remark this answers is not in the document]
Comment (myakove) [reply ocp_utilities/infra.py:307]:
```
with cluster_resource(Secret)(
        name="pull-secret", namespace="openshift-config"
    ) as pull_secret:
        ResourceEditor(patches={pull_secret: {"data": ps_content}}).update()
```

This code will not work.
You do not need to create the Secret

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_utilities/infra.py [hunk 2]
```diff
@@ -354,3 +357,32 @@ def create_icsp(icsp_name, repository_digest_mirrors):
     )
     icsp.deploy()
     return icsp
+
+
+def update_pull_secret(new_content_dict, secret_name, secret_namespace):
```
Comment (myakove) [inline ocp_utilities/infra.py:362]: Please add docstring and have some example how `new_content_dict` should look

Comment (myakove) [inline ocp_utilities/infra.py:362]: the name is update_ ... but you will create it if not exists, this needs to be rename

File: ocp_utilities/infra.py [hunk 3]
```diff
@@ -354,3 +357,32 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+    if secret.exists:
+        secret_data_dict = json.loads(
+            base64.b64decode(secret.instance.data[".dockerconfigjson"])
+        )["auths"]
+        secret_data_dict.update(new_content_dict["auths"])
+        secret_data_dict_encoded = base64.b64encode(
+            str({"auths": secret_data_dict}).replace("'", '"').encode("ascii")
+        ).decode("utf-8")
+
+        ResourceEditor(
```
Comment (myakove) [inline ocp_utilities/infra.py:374]:
```
ResourceEditor(....
return secret
```
then you can remove the `else:`

File: ocp_utilities/infra.py [hunk 4]
```diff
@@ -354,3 +357,32 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+            patches={secret: {"data": {".dockerconfigjson": secret_data_dict_encoded}}}
+        ).update()
+    else:
+        secret_data_encoded = base64.b64encode(
+            str(new_content_dict).replace("'", '"').encode("ascii")
+        ).decode("utf-8")
+        with cluster_resource(Secret)(
```
Comment (myakove) [inline ocp_utilities/infra.py:381]: the secret will be deleted once you exit this function.

File: ocp_utilities/infra.py [hunk 5]
```diff
@@ -354,3 +357,32 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+            name=secret_name,
+            namespace=secret_namespace,
+            data_dict={".dockerconfigjson": secret_data_encoded},
+        ) as secret:
+            secret.deploy()
```
Comment (myakove) [inline ocp_utilities/infra.py:386]: `with` already call `secret.deploy()`

File: ocp_utilities/infra.py [hunk 6]
```diff
@@ -354,3 +357,32 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+        secret_data_dict_encoded = base64.b64encode(
+            str({"auths": secret_data_dict}).replace("'", '"').encode("ascii")
+        ).decode("utf-8")
+
+        ResourceEditor(
+            patches={secret: {"data": {".dockerconfigjson": secret_data_dict_encoded}}}
+        ).update()
+    else:
+        secret_data_encoded = base64.b64encode(
+            str(new_content_dict).replace("'", '"').encode("ascii")
```
Comment (myakove) [inline ocp_utilities/infra.py:379]: why `("'", '"')` is needed?

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_utilities/infra.py (hunk 2 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

Comment (omrirh) [reply ocp_utilities/infra.py]: `create_update_pull_secret` maybe @myakove ?

File: ocp_utilities/infra.py (hunk 3 above)
Comment (omrirh) [reply ocp_utilities/infra.py:396]: done

File: ocp_utilities/infra.py (hunk 4 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: removed context manager to create the pull_secret resource

File: ocp_utilities/infra.py (hunk 5 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: removed context manager

File: ocp_utilities/infra.py (hunk 6 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: replaced with `json.dumps` ; the reason was that python dicts may use single/double quotes for key values while json pattern (which .dockerconfigjson value supports for encoding) only accepts double quotes. using `replace` was a bad w/a

**Review by myakove — CHANGES_REQUESTED** [review]

**Comment by myakove** [comment]
Please fix sonar issue
https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=193&id=RedHatQE_openshift-python-utilities&open=AYgA38t50TeVprt9rTMN

File: ocp_utilities/infra.py [hunk 7]
```diff
@@ -354,3 +357,56 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+                    <registry_name>:
+                        {"auth": <auth_token>,
+                        "email": <auth_email>}
+                    }
+                }
+        ps_name (str): Pull-secret name
+        ps_namespace (str): Pull-secret namespace
+
+    Returns:
+        pull_secret: A new pull-secret with new_content_dict as secret data if
```
[the remark this answers is not in the document]
Comment (rnetser) [reply ocp_utilities/infra.py:381]: should be updated in all places (pull secret is just a secret)

File: ocp_utilities/infra.py [hunk 8]
```diff
@@ -354,3 +357,56 @@ def create_icsp(icsp_name, repository_digest_mirrors):
     )
     icsp.deploy()
     return icsp
+
+
+def create_update_pull_secret(new_content_dict, ps_name, ps_namespace):
+    """Updating existing pull-secrets / Creating new pull-secrets
+
+    Args:
+        new_content_dict (dict): Secret data to be added/created,
```
Comment (rnetser) [inline ocp_utilities/infra.py:366]:
```suggestion
       secret_data_dict (dict): Secret data to be added/created,
```

Comment (rnetser) [inline ocp_utilities/infra.py:366]:
```suggestion
       secret_data_dict (dict): Secret data to be added/created,
```

File: ocp_utilities/infra.py [hunk 9]
```diff
@@ -354,3 +357,56 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+    pull_secret = Secret(name=ps_name, namespace=ps_namespace)
+
+    if pull_secret.exists:
+        ps_data_dict = json.loads(
+            base64.b64decode(pull_secret.instance.data[".dockerconfigjson"])
+        )["auths"]
+        ps_data_dict.update(new_content_dict["auths"])
+        ps_data_dict_encoded = base64.b64encode(
+            json.dumps({"auths": ps_data_dict}).encode("ascii")
+        ).decode("utf-8")
```
Comment (rnetser) [inline ocp_utilities/infra.py:394]: duplicate code; please extract to a function.

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_utilities/infra.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (omrim12) [reply ocp_utilities/infra.py:381]: This function is dedicated for pull-secrets (as the target is being modified under `.dockerconfigjson` value).
Updating the returned value

File: ocp_utilities/infra.py [hunk 10]
```diff
@@ -354,3 +357,56 @@ def create_icsp(icsp_name, repository_digest_mirrors):
     )
     icsp.deploy()
     return icsp
+
+
+def create_update_pull_secret(new_content_dict, ps_name, ps_namespace):
```
[the remark this answers is not in the document]
Comment (omrim12) [reply ocp_utilities/infra.py]: done

File: ocp_utilities/infra.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (omrirh) [reply ocp_utilities/infra.py]: This function is dedicated for pull-secrets (as the target secret is being modified under `.dockerconfigjson` value).
Updating the returned value

File: ocp_utilities/infra.py (hunk 8 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

Comment (omrirh) [reply ocp_utilities/infra.py]: done

File: ocp_utilities/infra.py (hunk 9 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

File: ocp_utilities/infra.py [hunk 11]
```diff
@@ -354,3 +357,69 @@ def create_icsp(icsp_name, repository_digest_mirrors):
     )
     icsp.deploy()
     return icsp
+
+
+def encode_secret_data(secret_data_dict):
```
Comment (rnetser) [inline ocp_utilities/infra.py:362]:
```suggestion
def dict_base64_encode(_dict):
```

Comment (rnetser) [reply ocp_utilities/infra.py:362]: remove the word `secret`; not relevant in this function

File: ocp_utilities/infra.py [hunk 12]
```diff
@@ -354,3 +357,69 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+    Returns:
+        secret_data_encoded (str): secret_data_dict encoded in base64
+    """
+    return base64.b64encode(json.dumps(secret_data_dict).encode("ascii")).decode(
+        "utf-8"
+    )
+
+
+def create_update_pull_secret(secret_data_dict, name, namespace):
+    """Updating existing pull-secret / Creating new pull-secret
```
Comment (rnetser) [inline ocp_utilities/infra.py:376]:
```suggestion
    """Update existing pull-secret or create a new pull-secret
```

File: ocp_utilities/infra.py [hunk 13]
```diff
@@ -354,3 +357,69 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+
+    Args:
+        secret_data_dict (dict): Secret data to be added/created,
+            e.g.: {"auths":{
```
Comment (rnetser) [inline ocp_utilities/infra.py:380]:
```suggestion
            example: {"auths":{
```

File: ocp_utilities/infra.py [hunk 14]
```diff
@@ -354,3 +357,69 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+                        "email": <auth_email>}
+                    }
+                }
+        name (str): Secret name
+        namespace (str): Secret namespace
+
+    Returns:
+        pull_secret (Secret): A new pull-secret with secret_data_dict as secret data if
+            the pull-secret does not exist, else an updated secret with
+            secret_data_dict data addition is returned.
```
Comment (rnetser) [inline ocp_utilities/infra.py:396]:
```suggestion
       Secret object
```

File: ocp_utilities/infra.py [hunk 15]
```diff
@@ -354,3 +357,69 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+    """
+    pull_secret = Secret(name=name, namespace=namespace)
```
Comment (rnetser) [inline ocp_utilities/infra.py:398]:
```suggestion
    secret = Secret(name=name, namespace=namespace)
```

File: ocp_utilities/infra.py (hunk 12 above)
Comment (rnetser) [inline ocp_utilities/infra.py:376]:
```suggestion
    """Update existing secret or create a new secret; secret type - dockerconfigjson
```
please remove "pull" - this is not relevant. you work with `secret`

**Review by rnetser — CHANGES_REQUESTED** [review]
please remove the word `pull` and add a ref to `dockerconfigjson` - this is the secret type that this code handles

**Review by rnetser — CHANGES_REQUESTED** [review]
please remove the word `pull` and add a ref to `dockerconfigjson` - this is the secret type that this code handles

File: ocp_utilities/infra.py (hunk 11 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

File: ocp_utilities/infra.py (hunk 12 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

File: ocp_utilities/infra.py (hunk 13 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

File: ocp_utilities/infra.py (hunk 14 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

File: ocp_utilities/infra.py (hunk 15 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

File: ocp_utilities/infra.py (hunk 12 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

**Review by rnetser — CHANGES_REQUESTED** [review]
Please also update the commit title (remove `pull`); this goes into the release log

**Review by rnetser — DISMISSED** [review]

File: ocp_utilities/infra.py [hunk 16]
```diff
@@ -354,3 +357,65 @@ def create_icsp(icsp_name, repository_digest_mirrors):
     )
     icsp.deploy()
     return icsp
+
+
+def dict_base64_encode(_dict):
+    """Encoding dict in base64
```
Comment (myakove) [inline ocp_utilities/infra.py:363]: please add new line after `"""` (For all docstrings)

File: ocp_utilities/infra.py [hunk 17]
```diff
@@ -354,3 +357,65 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+
+    Args:
+        _dict (dict): data dict to be encoded
+    Returns:
```
Comment (myakove) [inline ocp_utilities/infra.py:367]: Empty ine before `Returns`

File: ocp_utilities/infra.py [hunk 18]
```diff
@@ -354,3 +357,65 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+        _dict encoded in base64 (str)
```
Comment (myakove) [inline ocp_utilities/infra.py:368]: syntax is: `type: description` (For all docstrings)

File: ocp_utilities/infra.py [hunk 19]
```diff
@@ -354,3 +357,65 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+        )
+
+        ResourceEditor(
+            patches={secret: {"data": {secret_key: secret_data_encoded}}}
+        ).update()
+
+        return secret
+
+    secret_data_encoded = dict_base64_encode(_dict=secret_data_dict)
+    secret = Secret(
```
Comment (myakove) [inline ocp_utilities/infra.py:414]: You already have Secret object defined in line 394

Comment (myakove) [inline ocp_utilities/infra.py:414]: return Secret(....)

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_utilities/infra.py (hunk 16 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: done

File: ocp_utilities/infra.py (hunk 17 above)
Comment (omrirh) [reply ocp_utilities/infra.py:369]: done

File: ocp_utilities/infra.py (hunk 19 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: applying `secret.res.update(...)` and `secret.deploy` instead of old method

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_utilities/infra.py [hunk 20]
```diff
@@ -354,3 +357,64 @@ def create_icsp(icsp_name, repository_digest_mirrors):
...
+        )
+
+        ResourceEditor(
+            patches={secret: {"data": {secret_key: secret_data_encoded}}}
+        ).update()
+
+        return secret
+
+    secret_data_encoded = dict_base64_encode(_dict=secret_data_dict)
+    secret.res.update({"data": {secret_key: secret_data_encoded}})
```
Comment (rnetser) [inline ocp_utilities/infra.py:417]: secret.data_dict = ...

**Review by rnetser — CHANGES_REQUESTED** [review]
see above comments

File: ocp_utilities/infra.py (hunk 20 above)
Comment (omrirh) [reply ocp_utilities/infra.py:417]: Getting an error when trying to modify `secret.data_dict` directly like this:
```
secret.data_dict = {"data": {secret_key: secret_data_encoded}}
```

The http response:
```
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Secret in version \\"v1\\" cannot be handled as a Secret: json: cannot unmarshal object into Go struct field Secret.data of type []uint8","reason":"BadRequest","code":400}

Comment (rnetser) [reply ocp_utilities/infra.py:417]: check the structure of what you are trying to create; you have an unnecessary level

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_utilities/infra.py (hunk 20 above)
Comment (omrirh) [reply ocp_utilities/infra.py]: right (:
missed that.
done

**Review by rnetser — APPROVED** [review]

**Review by myakove — APPROVED** [review]

**Pull request merged**